engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/ConstructorBranchNode.java

```diff
@@ -40,38 +42,44 @@ public static ConstructorBranchNode build(
 
   @Specialization
   void doAtom(
-      VirtualFrame frame, Object state, Atom target, @Cached FieldsAsArrayNode toArrayNode) {
+      VirtualFrame frame,
+      Object state,
+      Atom target,
+      @Shared @CachedLibrary(limit = "3") StructsLibrary structsLib) {
     if (profile.profile(matcher == target.getConstructor())) {
-      var arr = toArrayNode.executeFields(target);
+      var arr = fieldsFromObject(target, matcher, structsLib);
+      accept(frame, state, arr);
+    }
+  }
+
+  @Specialization
+  void doMultiValue(
+      VirtualFrame frame,
+      Object state,
+      EnsoMultiValue target,
+      @Cached IsValueOfTypeNode isValueOfTypeNode,
+      @Shared @CachedLibrary(limit = "3") StructsLibrary structsLib,
+      @Cached EnsoMultiValue.CastToNode castNode) {
+    var expectedType = matcher.getType();
+    if (profile.profile(isValueOfTypeNode.execute(expectedType, target, true))) {
+      var replacement = castNode.findTypeOrNull(expectedType, target, true, true);
```

Comment:
```suggestion
      var replacement = castNode.findTypeOrNull(expectedType, target, true, false);
```
I think we want to ignore hidden types

Reply:
Done in 56e79ee1f992b84b7a87897ec362edfc5bf43cae